### DISCUSSION TITLE
Alignment and Intensity differences between eyepy extraction and constructor viewer

### DISCUSSION THREAD

**Discussion by MathiasGallardo** [MedVisBonn/eyepy#19] [Q&A] [answered]
Dear Eyepy community,

I observe that the saved Bscans from the e2e file (OCT volume) using eyepy present some differences with the ones visible in the Heidelberg viewer.
Here is the code snippet I used for the eyepy extraction:
`# Import HEYEX E2E export
    ev = ep.import_heyex_e2e(path_e2e)
    # Save as PNG
    for i in range(0,len(ev)):
        imageio.imwrite(f'bscan_{i}.png', ev[i].data)`

I noticed that the intensities are not the same as those of the Bscans visible in the viewer. In my case, I have the impression that the Bscans are lighter in the viewer.

Also, the retina is not placed at the same height of the Bscans along the volume.

Does anyone have encountered these differences?
Also, any solution to be aligned with the Bscans in the viewer?

Many thanks in advance,

**Comment by Oli4** [comment, marked as answer]
Dear @MathiasGallardo,

The code you're using to read and save the B-scans looks good. Unfortunately, I haven't fully figured out the E2E format, and I'm not aware of anyone who has. The E2E format stores raw intensities, which differ from what is displayed in the Heyex viewer. I once ran some [experiments](https://github.com/marksgraham/OCT-Converter/issues/21#issuecomment-1057455183) to understand the intensity transform used in the XML export, but it's possible that the viewer uses a different or updated transformation.

If you'd prefer a different intensity transformation, you can define your own function to modify the raw intensities:

```python
# Load your EyeVolume object (assuming `ev` is your EyeVolume object)
ev = EyeVolume(...)  # Load your data here

# Set your custom intensity transform function
ev.set_intensity_transform(custom_intensity_transform)

# Access the transformed data
transformed_data = ev.data
```

Additionally, the B-scans in the E2E files are not registered. There might be an affine matrix somewhere in the E2E file, but I haven't found it yet. I've wanted to add registration functionality to eyepy for cases like this, but I haven't had the time to integrate it.

You're very welcome to add such functionality, and I'd be happy to review and accept a related pull request.

Best,
Olivier
